### PR TITLE
[SPARK-17440] [SPARK-17441] Fixed Multiple Bugs in ALTER TABLE

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

For the following `ALTER TABLE` DDL, we should issue an exception when the target table is a `VIEW`:

``` SQL
 ALTER TABLE viewName SET LOCATION '/path/to/your/lovely/heart'

 ALTER TABLE viewName SET SERDE 'whatever'

 ALTER TABLE viewName SET SERDEPROPERTIES ('x' = 'y')

 ALTER TABLE viewName PARTITION (a=1, b=2) SET SERDEPROPERTIES ('x' = 'y')

 ALTER TABLE viewName ADD IF NOT EXISTS PARTITION (a='4', b='8')

 ALTER TABLE viewName DROP IF EXISTS PARTITION (a='2')

 ALTER TABLE viewName RECOVER PARTITIONS

 ALTER TABLE viewName PARTITION (a='1', b='q') RENAME TO PARTITION (a='100', b='p')
```

In addition, `ALTER TABLE RENAME PARTITION` is unable to handle data source tables, just like the other `ALTER PARTITION` commands. We should issue an exception instead. 
### How was this patch tested?

Added a few test cases.

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @cloud-fan @yhuai

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 1]
```diff
@@ -318,44 +331,51 @@ class HiveDDLSuite
...
+
+        checkMisuseForAlterTableOrView(
+          s"$oldViewName SET SERDEPROPERTIES ('x' = 'y')", isAlterView = false)
+
+        checkMisuseForAlterTableOrView(
+          s"$oldViewName PARTITION (a=1, b=2) SET SERDEPROPERTIES ('x' = 'y')", isAlterView = false)
+
+        checkMisuseForAlterTableOrView(
+          s"$oldViewName ADD IF NOT EXISTS PARTITION (a='4', b='8')", isAlterView = false)
+
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: We also need to test `isAlterView` = `true` for `ADD PARTITION` and `DROP PARTITION`. Let me add them now...

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: uh, both are not supported. nvm. I think we already cover all the cases. Code is ready to review. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 2]
```diff
@@ -693,14 +705,22 @@ object DDLUtils {
       catalog: SessionCatalog,
       tableIdentifier: TableIdentifier,
       isView: Boolean): Unit = {
-    catalog.getTableMetadataOption(tableIdentifier).map(_.tableType match {
+    catalog.getTableMetadataOption(tableIdentifier).foreach(
+      verifyAlterTableType(_, tableIdentifier, isView))
+  }
+
+  def verifyAlterTableType(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: can we combine these 2 `verifyAlterTableType` into one?

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: I think we only need this one, and we can remove the old one.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 3]
```diff
@@ -383,7 +385,14 @@ case class AlterTableRenamePartitionCommand(
   extends RunnableCommand {
 
   override def run(sparkSession: SparkSession): Seq[Row] = {
-    sparkSession.sessionState.catalog.renamePartitions(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: what's the behaviour before?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 4]
```diff
@@ -881,23 +889,30 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
...
-      Set(Map("a" -> "100", "b" -> "p"), Map("a" -> "200", "b" -> "c"), part3))
-    // rename without explicitly specifying database
-    catalog.setCurrentDatabase("dbx")
-    sql("ALTER TABLE tab1 PARTITION (a='100', b='p') RENAME TO PARTITION (a='10', b='p')")
-    assert(catalog.listPartitions(tableIdent).map(_.spec).toSet ==
-      Set(Map("a" -> "10", "b" -> "p"), Map("a" -> "200", "b" -> "c"), part3))
+    if (isDatasourceTable) {
+      convertToDatasourceTable(catalog, tableIdent)
+    }
+    maybeWrapException(isDatasourceTable) {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: nit: `maybeWrapException(expectException = isDatasourceTable)`

Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: can we also test the error message?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 5]
```diff
@@ -869,6 +869,14 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
...
 
   test("alter table: rename partition") {
+    testRenamePartition(isDatasourceTable = false)
+  }
+
+  test("alter table: rename partition (datasource table)") {
+    testRenamePartition(isDatasourceTable = true)
+  }
+
+  private def testRenamePartition(isDatasourceTable: Boolean): Unit = {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: I'd like to create a method to set up environment to test rename partition, instead of creating a common testing method and pass in a flag.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 6]
```diff
@@ -305,6 +305,19 @@ class HiveDDLSuite
     }
   }
 
+  private def checkMisuseForAlterTableOrView(sqlText: String, isAlterView: Boolean): Unit = {
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: I'd like to create a `assertErrorForAlterTableOnView`, which only checks ALTER TABLE when the table is a view. For ALTER VIEW when table is not a view, we can create a new method, or just duplicate the code if the test cases are not many.

**Comment by gatorsmile** [comment]
Accepted all the comments. Will address comments now. : )

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: We got the following errors:
```
Partition not found in table 'tab1' database 'default':
ds -> 2008-04-08
hr -> 11;
org.apache.spark.sql.catalyst.analysis.NoSuchPartitionException: Partition not found in table 'tab1' database 'default':
ds -> 2008-04-08
hr -> 11;
	at org.apache.spark.sql.hive.client.HiveClientImpl$$anonfun$renamePartitions$1$$anonfun$apply$mcV$sp$5$$anonfun$16.apply(HiveClientImpl.scala:502)
	at org.apache.spark.sql.hive.client.HiveClientImpl$$anonfun$renamePartitions$1$$anonfun$apply$mcV$sp$5$$anonfun$16.apply(HiveClientImpl.scala:502)
```

Basically, Hive does not recognize the partitioning info of data source tables.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 7]
```diff
@@ -871,32 +871,55 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
...
-    createTable(catalog, tableIdent)
-    createTablePartition(catalog, part1, tableIdent)
-    createTablePartition(catalog, part2, tableIdent)
-    createTablePartition(catalog, part3, tableIdent)
-    assert(catalog.listPartitions(tableIdent).map(_.spec).toSet ==
-      Set(part1, part2, part3))
+    createPartitionedTable(tableIdent, isDatasourceTable = false)
     sql("ALTER TABLE dbx.tab1 PARTITION (a='1', b='q') RENAME TO PARTITION (a='100', b='p')")
-    sql("ALTER TABLE dbx.tab1 PARTITION (a='2', b='c') RENAME TO PARTITION (a='200', b='c')")
+    sql("ALTER TABLE dbx.tab1 PARTITION (a='2', b='c') RENAME TO PARTITION (a='20', b='c')")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: unnecessary change?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 8]
```diff
@@ -871,32 +871,55 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
...
+      Set(Map("a" -> "10", "b" -> "p"), Map("a" -> "20", "b" -> "c"), Map("a" -> "3", "b" -> "p")))
+    // partition to rename does not exist
+    intercept[NoSuchPartitionException] {
+      sql("ALTER TABLE tab1 PARTITION (a='not_found', b='1') RENAME TO PARTITION (a='1', b='2')")
+    }
     // table to alter does not exist
     intercept[NoSuchTableException] {
       sql("ALTER TABLE does_not_exist PARTITION (c='3') RENAME TO PARTITION (c='333')")
     }
-    // partition to rename does not exist
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: nit: why change the test order? I think it's natural to test non-existing table first, then non-existing partition.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 9]
```diff
@@ -1246,7 +1269,7 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
     }
     // table to alter does not exist
     intercept[AnalysisException] {
-      sql("ALTER TABLE does_not_exist SET SERDEPROPERTIES ('x' = 'y')")
+      sql("ALTER TABLE does_not_exist PARTITION (a=1, b=2) SET SERDEPROPERTIES ('x' = 'y')")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: why change this test? If we wanna test the PARTITION code path, we should add a new test

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala (hunk 7 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: https://github.com/gatorsmile/spark/blob/5d0cffe6b59dc36d63f7d1e055903b53ce2fec42/sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala#L878 will be longer than 100 characters... That is why I changed it to avoid adding another line.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala (hunk 8 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: Yeah, let me change it back.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala (hunk 9 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: This is to fix a bug in `testSetSerdePartition`. This function is to test `ALTER TABLE PARTITION Set SERDEPROPERTIES`. It sounds like the previous coder just copied and pasted from another function.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala (hunk 7 above)
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: ok then it's fine...

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 10]
```diff
@@ -712,7 +712,7 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
         """.stripMargin)
 
       val e = intercept[AnalysisException] {
-        sql("ALTER TABLE tab1 RENAME TO tab2")
+        sql("ALTER VIEW tab1 RENAME TO tab2")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: Sorry I missed this one. So we can't use `ALTER TABLE` to alter temp view now?

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: en... You are right. Let me resolve this soon.

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: Will also add a few test cases for ensuring `ALTER TABLE` commands can alter temporary views.

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: The problem is fixed. : )

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 11]
```diff
@@ -687,20 +699,23 @@ object DDLUtils {
 
   /**
    * If the command ALTER VIEW is to alter a table or ALTER TABLE is to alter a view,
-   * issue an exception [[AnalysisException]].
+   * issue an exception [[AnalysisException]]. Temporary views can be altered by both
+   * ALTER VIEW and ALTER TABLE commands.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: we may also explain the reason, i.e. temporary views can be created by `CREATE TEMPORARY TABLE`

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: and we can remove this trick if we decide to not allow temporary views to be created by `CREATE TEMPORARY TABLE`.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: Yeah. Let me do it now. Thanks!

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: The comment is added. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 12]
```diff
@@ -383,7 +385,14 @@ case class AlterTableRenamePartitionCommand(
   extends RunnableCommand {
 
   override def run(sparkSession: SparkSession): Seq[Row] = {
-    sparkSession.sessionState.catalog.renamePartitions(
+    val catalog = sparkSession.sessionState.catalog
+    val table = catalog.getTableMetadata(tableName)
+    if (DDLUtils.isDatasourceTable(table)) {
+      throw new AnalysisException(
+        "ALTER TABLE RENAME PARTITION is not allowed for tables defined using the datasource API")
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: we can just say `data source table`, to be consistent with other places.

**Comment by cloud-fan** [comment]
thanks, merging to master! you can address my last comment in your other PRs

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala (hunk 12 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: Sure, will do it. Thanks!

**Comment by gatorsmile** [comment]
Thank you! : )
